src/main/java/com/iexec/core/replicate/ReplicatesService.java

```diff
@@ -303,6 +291,53 @@ public void updateReplicateStatus(String chainTaskId,
         }
 
         if (!canUpdate) {
+            return Optional.of(ReplicateStatusUpdateError.GENERIC_CANT_UPDATE);
+        }
+
+        return Optional.empty();
+    }
+
+    /*
+     * We retry up to 100 times in case the task has been modified between
+     * reading and writing it.
+     *
+     * Before updating we check:
+     *   1) if valid transition.
+     *   2) if worker did fail when CONTRIBUTE/REVEAL/UPLOAD_FAILED.
+     *   3) if worker did succeed onChain when CONTRIBUTED/REVEALED.
+     *   4) if worker did upload when RESULT_UPLOADING.
+     */
+    @Retryable(value = {OptimisticLockingFailureException.class}, maxAttempts = 100)
+    public Optional<TaskNotificationType> updateReplicateStatus(String chainTaskId,
+                                                                String walletAddress,
+                                                                ReplicateStatusUpdate statusUpdate) {
+        log.info("Replicate update request [status:{}, chainTaskId:{}, walletAddress:{}, details:{}]",
+                statusUpdate.getStatus(), chainTaskId, walletAddress, statusUpdate.getDetailsWithoutStdout());
+
+        final Optional<ReplicateStatusUpdateError> replicateStatusUpdateError =
+                canUpdateReplicateStatus(chainTaskId,walletAddress,statusUpdate);
```

Comment:
Why do we make this call two times here and in the controller?

Reply:
Some methods depends on `ReplicateService#updateReplicateStatus` so it should continue to check whether the replicate status could be updated while the controller must know why the status couldn't be updated.
I have found two solutions to this problem:

1. Reformat all the methods depending on `ReplicateService#updateReplicateStatus` so that it can update without checking whether it is a valid update;
2. Make sure everything still works by having twice the test.

First solution would be better but it may lead to new issues if code coverage is not wide. Would you still prefer to go for the first solution?

Reply:
I see. In `ReplicateService#updateReplicateStatus` when we have `ALREADY_REPORTED` can we return `ReplicateWorkflow.getInstance().getNextAction(...)` since "already reported" means the worker can proceed to the following steps of the workflow? In that case, we wouldn't need the verification in the controller. What do you think?

Reply:
I was thinking about this solution, but I'm not sure whether `ReplicateWorkflow.getInstance().getNextAction(...)` would return the correct next action. What if `core` is 2 steps behind `worker`'s status update? It would probably break something.
Do you have any input on this @jeremyjams?

Reply:
Core should ask the worker not to do anything else, since another process should already have notified the worker about the next real action.

Reply:
Which TaskNotificationType denotes "do not do anything else"? Is that "PLEASE_CONTINUE"?

Reply:
I guess an appropriate reply would be `PLEASE_WAIT` since "Please wait since you should already have received other orders" looks meaningful.

Reply:
So I have done a few tests on this point. When using `PLEASE_WAIT`, the worker seems to not do anything more:
```
2021-08-25 13:34:00.484  INFO 1 --- [         task-5] c.i.w.executor.TaskNotificationService   : update replicate response [chainTaskId:0x11f3f6938ff0edf43fc0a11dedb68b9d5802c3a7c261a49d0d74b49c6e48124a, status:CONTRIBUTED, next:PLEASE_WAIT]
2021-08-25 13:34:00.485  WARN 1 --- [         task-5] c.i.w.executor.TaskNotificationService   : [DEBUG] Action: PLEASE_CONTRIBUTE, next action: PLEASE_WAIT
2021-08-25 13:34:00.485  WARN 1 --- [         task-6] c.i.w.executor.TaskNotificationService   : [DEBUG] Action: PLEASE_WAIT, next action: null
2021-08-25 13:34:00.486  WARN 1 --- [         task-6] c.i.w.executor.TaskNotificationService   : No more actions to do [chainTaskId:0x11f3f6938ff0edf43fc0a11dedb68b9d5802c3a7c261a49d0d74b49c6e48124a]
```

Same goes for `PLEASE_CONTINUE`:
```
2021-08-25 13:41:42.224  INFO 1 --- [         task-5] c.i.w.executor.TaskNotificationService   : update replicate response [chainTaskId:0x553f0f4a6666dc1e1242b6f8c9ef7294e305aaf1e36bd942f656fda0e3dc3411, status:CONTRIBUTED, next:PLEASE_CONTINUE]
2021-08-25 13:41:42.224  WARN 1 --- [         task-5] c.i.w.executor.TaskNotificationService   : [DEBUG] Action: PLEASE_CONTRIBUTE, next action: PLEASE_CONTINUE
2021-08-25 13:41:42.225  WARN 1 --- [         task-6] c.i.w.executor.TaskNotificationService   : [DEBUG] Action: PLEASE_CONTINUE, next action: null
2021-08-25 13:41:42.225  WARN 1 --- [         task-6] c.i.w.executor.TaskNotificationService   : No more actions to do [chainTaskId:0x553f0f4a6666dc1e1242b6f8c9ef7294e305aaf1e36bd942f656fda0e3dc3411]
```

I'm not sure of the flow of the worker, however it seems to loop on the `TaskNotificationService#onTaskNotification` method as long as it has actions to do. There's no code to handle `PLEASE_WAIT`, `PLEASE_CONTINUE` or `null` so it seems logic to stop there.
I'll continue to search.